Implemented deriveInterpolatingValue function

pdyraga opened the pull request:
deriveInterpolatingValue implements `def derive_interpolating_value(L, x_i)` function from [FROST], as defined in section 4.2 Polynomials.

deriveInterpolatingValue is a simplified version of the algorithm for computing Lagrange coefficient given we only care about the value f(0) instead of reconstructing the entire polynomial.

`L` is the list of the indices of the members of the particular group. `xi` is the index of the participant `i`.

The function is identical as in @eth-r's prototype except that we return errors instead of panicking. Unit tests covering the function and explaining how interpolating value is calculated were added.

pdyraga committed:
Implemented deriveInterpolatingValue function

deriveInterpolatingValue implements def derive_interpolating_value(L, x_i)
function from [FROST], as defined in section 4.2 Polynomials.

deriveInterpolatingValue is a simplified version of the algorithm for
computing Lagrange coefficient given we only care about the value f(0) instead
of reconstructing the entire polynomial.

L is the list of the indices of the members of the particular group.
xi is the index of the participant i.

eth-r commented on frost/signer.go:
```diff
@@ -364,3 +364,87 @@ func (s *Signer) encodeGroupCommitment(commitments []*NonceCommitment) []byte {
+			// for x_j in L:
+			//     if count(x_j, L) > 1:
+			//         raise "invalid parameters"
+			if found {
```
This doesn't actually implement the specified functionality. We need to expect that the indices `L` are sorted in ascending order, with no repeated instances, and keep track of the previous `x_j'`. Comparing `x_j` to `x_j'` and requiring that `x_j > x_j'` gives the specified functionality. This was a genuine oversight in the prototype.

pdyraga replied:
That's true though looking at `Round2` implementation in #8 I do wonder if we need to worry about it in `deriveInterpolatingValue` at all. `validateGroupCommitments` makes sure that all commitments are sorted in ascending order and there is no duplicates.

We could improve unit tests for `validateGroupCommitments` to cover duplicates and make it clear in `deriveInterpolatingValue` that the validation must happen before this function is called.

This would allow us to eliminate this error handling:
https://github.com/threshold-network/roast-go/blob/579570232aa2bd85881be676eec5dd47dceade5f/frost/signer.go#L120-L125

and `validateGroupCommitments` would be the only function returning errors in `Round2`. This wouldn't be 1:1 with the paper but this solution feels quite elegant to me.

What do you think about it?

pdyraga replied:
Regardless of the decision here, I added a test case to `validateGroupCommitments` unit tests covering the duplicate element case: 1398d05c9b025927f46203b0f6f2608526946e97.

eth-r replied:
I think if we note in a comment that `validateGroupCommitments` is a necessary part for checking the correctness, it should be fine.

pdyraga replied:
I will handle it in #8.

eth-r commented on frost/ciphersuite.go:
```diff
@@ -46,6 +46,9 @@ type Curve interface {
 	Identity() *Point
 
+	// Order returns the order of the elliptic curve.
+	Order() *big.Int
```
For some potential future implementations, this comment will need to be amended to "returns the order of the group produced by the generator". On some curves the generator produces a prime-order subgroup, whose order is not the same as that of the entire curve.

pdyraga replied:
That's a really good point. Improved in e1c912ce48e7797a9e1470bca8cefd36b373c061.

eth-r reviewed (changes requested):
Found one real error stemming from the prototype implementation. Also one minor documentation detail.

pdyraga committed:
Improved the Order function documentation

On some curves the generator produces a prime-order subgroup, whose
order is not the same as that of the entire curve.

pdyraga committed:
Improved test coverage of validateGroupCommitments

Improved the test coverage to include the duplicate commitment case.

eth-r reviewed (approved):
Looks good.
